### PR TITLE
fix: handle decimal point for spinbox #1051

### LINKED ISSUE #1051: Manual input to spinbox partly broken
#### Wave SDK Version, OS

Wave 0.18, Ubuntu 18

#### Actual behavior

When trying to manually input data into a `spinbox`, I can observe unexpected behavior:

https://user-images.githubusercontent.com/1677826/136384964-72ba518f-c4ee-4c0a-bd13-94524c07dcdf.mp4

In this example, I want to manually enter `0.01`. A common procedure is to delete everything, and add the value manually.
So in this case, I delete the current value, and try to input the new value. However, it does not allow to enter the decimal value.
At that point it breaks.

#### Expected behavior

Spinbox should allow to properly provide manual input.

cc @goldentom42

### PR DESCRIPTION
**Pull request by aalencar**
* User should be able to enter decimal points and increment/decrement without rounding.

closes #1051

### CODE REVIEW

**Comment by aalencar** [comment]
@mturoci tests are passing and I tested manually with `step` of `0.0001` and both increment and decrement work as expected.

[Copied Fluent UI approach](https://github.com/microsoft/fluentui/blob/25435def33d60a9314ccdcf63a1ee37f47513c50/packages/react/src/components/SpinButton/SpinButton.base.tsx#L291) for `handleIncrement` and `handleDecrement`

For some reason, Table Component tests are failing but I haven't touched it.

image

**Comment by aalencar** [comment]
> How is it possible that it works. I would expect the value not to be updated within input since setVal({ val: String(newValue) }) is not called at all and the component is controlled.

`value` doesn't need to be updated because `11.` won't be parsed and it's the same as the previous value which was `11`. Same for `11.0` or `11.010`. I'm not sure if I answered your question.

**Comment by mturoci** [comment]
>value doesn't need to be updated because 11. won't be parsed and it's the same as the previous value which was 11. Same for 11.0 or 11.010. I'm not sure if I answered your question.

Not really. I mean the value that appears in input - it somehow correctly gets updated to`11.` despite prev value is `11`. I get the part why we are not parsing it (source of bug) but I would expect this to make it work:
```ts
if (/\.$|\.\d*0+$/.test(e.target.value)) {
setVal({ val: '11.' }) // If we dont do this, the component should keep rendering old value of `11` instead of `11.` - wondering how can it be working :)
return
}

**Comment by aalencar** [comment]
> Just checked current master and they pass. Rebase against current master, stash your changes and see if it helps. If yes, the problem is in your code.

It also fails when I'm running the tests on `master` branch with the latest changes. I suspect that it has something to do with locale since the tests are related to time and date.

image

**Comment by aalencar** [comment]
> or 0 as the last number - is this really necessary? Our precision-handling logic should take care of that, right?

Yes, but we want to wait so that the user can enter a value after the `0` e.g. 11 -> 11. -> 11.0 -> 11.02
If we don't do this when the user types 11.0 the parse will bring it back to 11

File: ui/src/spinbox.tsx [hunk 1]
```diff
@@ -98,6 +102,15 @@ export const
...
       },
+      // Does what precisionRound does but for strings
+      trimIfOverPrecision = (newValue: S) => {
+        const [head, tail] = newValue.split('.')
+        return `${head}.${tail.slice(0, precision)}`
+      },
+      skipParsing = (newValue: S) => {
+        const isLastCharDotOrTraillingZero = /\.$|\.\d*0+$/
+        return precision > 0 && isLastCharDotOrTraillingZero.test(newValue)
+      },
```
[the remark this answers is not in the document]
Comment (aalencar) [reply ui/src/spinbox.tsx]: `trimIfOverPrecision` logic moved to where it's being used.

File: ui/src/table.test.tsx [hunk 2]
```diff
@@ -49,13 +49,14 @@ describe('Table.tsx', () => {
...
+    const date = '1971-07-08T23:09:33'
     tableProps = {
       ...tableProps,
-      rows: [{ name: 'rowname1', cells: ['1971-07-08T23:09:33'] }],
+      rows: [{ name: 'rowname1', cells: [date] }],
       columns: [{ name: 'colname1', label: 'Col1', sortable: true, searchable: true, data_type: 'time' }]
     }
     const { getAllByRole } = render(<XTable model={tableProps} />)
-    expect(getAllByRole('gridcell')[0].textContent).toBe('7/8/1971, 11:09:33 PM')
+    expect(getAllByRole('gridcell')[0].textContent).toBe(new Date(date).toLocaleString())
```
[the remark this answers is not in the document]
Comment (aalencar) [reply ui/src/table.test.tsx]: Fixed with env variable

File: ui/src/spinbox.test.tsx [hunk 3]
```diff
@@ -162,4 +162,13 @@ describe('Spinbox.tsx', () => {
...
   })
+
+  it('Should trim the value if overflows the step precision', () => {
+    const { getByTestId, container } = render(<XSpinbox model={{ ...spinboxProps, value: 10, step: 0.0001 }} />),
+      spinboxInput = container.querySelector('.ms-spinButton-input') as HTMLInputElement
+    
+    fireEvent.change(getByTestId(name), { target: { value: 0.000200001 } })
+
+    expect(spinboxInput.value).toBe('0.0002')
+  })
```
[the remark this answers is not in the document]
Comment (aalencar) [reply ui/src/spinbox.test.tsx]: Added tests

**Review by mturoci — APPROVED** [review]
Thanks @aalencar!

Added a few small changes:

* managed to fix the Fluent bug related to clicking increment / decrement in Fluent v7
* your newly added tests still didn't reach your newly added condition since you used numbers instead of strings so
```ts
fireEvent.change(spinboxInput, { target: { value: 0.010000000 } })
expect(wave.args[name]).toBe(0.01)
```
should be 
```ts
fireEvent.change(spinboxInput, { target: { value: '0.010000000' } })
expect(wave.args[name]).toBe(0.01)
```
as number `0.01000000` would be converted to `0.01`, but string would not. This broke the tests as `wave.args` were not correctly set for this case so needed fixing as well.

**Comment by aalencar** [comment]
@mturoci thanks for fixing the tests and the increment issue 👍 

The PR looks good to be merged

**Pull request merged**